FWCore/ServiceRegistry/interface/PathsAndConsumesOfModulesBase.h

```diff
@@ -66,17 +75,72 @@ namespace edm {
+    // although the elements contain a member that indicates the branch type
+    // for ED products and the transition associated with the consumes call
+    // for EventSetup products.
+    // Note the other functions above return a reference to
+    // an object that is held in memory throughout the job, while the
+    // following function returns a newly created object each time.  We
+    // do not expect this to be called during a normal production job where
     // performance and memory are important. These objects are bigger
     // than just a pointer.
     std::vector<ConsumesInfo> consumesInfo(unsigned int moduleID) const { return doConsumesInfo(moduleID); }
+    std::vector<EventSetupConsumesInfo> eventSetupConsumesInfo(unsigned int moduleID) const {
+      return doEventSetupConsumesInfo(moduleID);
+    }
 
     unsigned int largestModuleID() const { return doLargestModuleID(); }
 
+    // The returned vector contains an entry for every EventSetup module.
+    std::vector<eventsetup::ESProductResolverProvider const*> const& allESProductResolverProviders() const {
+      return doAllESProductResolverProviders();
+    }
+
+    // The returned container indicates which EventSetup modules depend on which other
+    // EventSetup modules. Each entry in the outer vector corresponds to an entry
+    // in the vector returned by allESProductResolverProviders (same size and order).
+    // The inner vector will contain an entry for each EventSetup module that can produce
+    // a consumed data product. In the case of "may consumes", the producers for
+    // all the products that might be consumed are included. The inner vector
+    // includes producers for all transitions together in the same vector and
+    // has had duplicates removed.
+    std::vector<std::vector<eventsetup::ComponentDescription const*>> const&
+    esModulesWhoseProductsAreConsumedByESModule() const {
+      return doESModulesWhoseProductsAreConsumedByESModule();
+    }
+
+    // Returns a container of information for one ESProducer. The outer vector is
+    // indexed by the transitionID, which is a counter of setWhatProduced calls in
+    // the order they occur in the ESProducer constructor. The inner vector has
+    // elements that correspond to the individual calls to the consumes function.
+    // In an individual element, an empty module type indicates there is no
+    // ESProducer or ESSource to produce the consumed data.
+    // There is also a flag to indicate a module labeled mismatch.
+    // "May consumes" elements are different because there can be multiple entries
+    // for one "may consumes" function call, one for each available product that
+    // matches the EventSetupRecordKey type and product type. If there are no
+    // matches there is still one entry with a flag indicating there were
+    // no matches.
+    std::vector<std::vector<EventSetupConsumesInfo>> eventSetupConsumesInfo(ESProducer const& esProducer) const {
```

Comment:
I have a little bit uneasy feeling of adding a dependence on `ESProducer` type here (although it is only forward-declared).

With EDModules I suppose we deal with these dependencies by passing in the module ID (that is just and integer) in this abstract interface, whereas the implementation (in `FWCore/Framework`) gets access to the actual worker/module objects via `Schedule`. I wonder if something similar would be feasible for ESProducers? (even for possible subsequent work if would be too much for this PR)

Reply:
This answer replies to this comment and several others above. I have not pushed it yet, but when I do the dependence on EventSetupProvider is removed from PathsAndConsumesOfModulesBase, EDConsumerBase, and ESProducer. It remains in PathsAndConsumesOfModules, but that is class derived from the abstract base and only accessible from EventProcessor and SubProcess. Took some work, but it is better thanks for that comment.